Allow a babelConfig option

lemonmade opened the pull request:
Fixes https://github.com/kulshekhar/ts-jest/issues/284

This PR allows the user to pass a `jest.globals.babelConfig` option, which should be an object with a full babel config that gets merged into the "necessary" bits of the config that were already being added.

Change itself is pretty small, but I really wasn't sure what to do for the tests. In the end, I mostly copied the `babelrc` tests, but there are some cases (most notably, merging of the plugins/ presets arrays) that are not addressed by these tests. Would love suggestions!

@GeeWee @kulshekhar let me know if there is someone else I should ping for reviews.

kulshekhar commented:
How does this behave if someone adds a config in the jest section AND has a `.babelrc`?

Also, please update the authors file with your name :)

morajabi commented:
I think it should override them all?

GeeWee commented:
I'm sorry I've been busy. I'll try to have a look at this soon.

GeeWee commented:
It does look pretty clean. I'm surprisingly okay with this I think, if the readme is updated.
I'll also like to know how it handles the case @kulshekhar describes with the .babelrc.

lemonmade commented:
@GeeWee @kulshekhar thanks for the feedback. I updated the README and AUTHORS files. As @morajabi, any options you manually pass in the config are overwritten by any matching values in the `.babelrc` file that is found (this is, I guess, Babel's way of merging configs).

Let me know if you'd like some more changes!

kulshekhar commented:
> any options you manually pass in the config are overwritten by any matching values in the .babelrc file that is found (this is, I guess, Babel's way of merging configs).

I'd like two things related to this:

- The readme updated with this information
- a test that addresses this behavior (so that if something changes in babel down the road, we can proactively address it)

Other than this, I'm following @GeeWee's lead on this PR

lemonmade commented:
Sorry for the long delay in getting back to this. I've updated the README with more details on the config merging, and added the only tests I could think of that would help to demonstrate what's going on. Let me know if they make sense to you, too.

morajabi commented:
I think `babelConfig` should override `.babelrc`. In most cases, we have one main `.babelrc` in the project for the main build, but we want to override some part of it with the custom `babelConfig` specificly for jest (IMO). Any thoughts on this?

lemonmade commented:
@morajabi I don't think there is a way to achieve this with the way Babel resolves configs; if babelrc is allowed, it seems to take precedence.

kulshekhar commented:
> I think babelConfig should override `.babelrc`

What I gathered from @lemonmade's comment above is that this is handled by babel and isn't something that the PR does. If that understanding is right, I don't think we should change what would be expected babel behaviour

Other than that, I think this PR is in good shape and can be merged in once @GeeWee gives the green light

GeeWee commented:
(Just out of curiousity @lemonmade - what kind of tool are you building?)

lemonmade commented:
Thanks for working through this with me, @geewee! The tool is just a wrapper around other tools like Webpack, Jest, ESLint, TypeScript, etc, that provides commands for build, test, lint, and more. It looks at the structure of the project and runs the correct commands with the correct config, allowing developers to basically have no-configuration access to best in class dev tooling.

GeeWee commented:
Sounds cool! If it's public, definitely shoot me a mail or a tweet once it's done!

yesmeck commented:
This is cool, when will this feature be released?

GeeWee commented:
I guess we could release it @kulshekhar ?

kulshekhar commented:
Right away after the PR with the version bump is approved :)

kulshekhar commented:
done
